### ISSUE TITLE
Fitting an AR model on non-continuous multi-trial data

### ISSUE THREAD

**Issue by gabrielibagon** [statsmodels/statsmodels#4145]
I have a univariate 2D dataset made up of many non-consecutive trials (or slices/windows) of a time series (one trial is not continuous with another trial). I would like to fit an AR model to this collection of trials for forecasting on an exogenous trial.

My first approach is to concatenate all of the trials into a long 1-dimensional array, padding with NaNs to separate the trials, and then use `AR(missing='drop')` to ignore the NaNs. However, I'm not sure how the `missing` parameter works - does "drop" simply remove the NaNs and concatenate the array on both sides? This is not what I would want, since the AR model would then be trained on the data as if it was continuous, which it is not. The points between the trials shouldn't have any correlation, and the model should not use data from one trial to predict the time series of a new trial.

In summary, is there a better way to train a univariate AR model on a collection of discontinuous windows of data? Thanks!

**Comment by josef-pkt** [comment]
The older tsa models AR/ARMA/ARIMA always assume no missing or fixed time differences and cannot handle this. The new statespace models like SARIMAX can handle missing observations.
http://www.statsmodels.org/devel/examples/notebooks/generated/statespace_sarimax_internet.html
@ChadFulton What's the best example, explanation how to do this?


On the other hand, for the estimation of the parameters in the AR case it is also possible to us OLS.
In this case we can create the lag matrix of the endog as exog, wo we have
`y_t = [y_{t-1}, ... y_{t-p}, constant, trend, ...] * params + error`

Then, each row can be treated as a single observation. The standard OLS missing='drop' handling will drop these observations that have a missing/nan in any of the y_t to y_{t-p} positions.

For stacking experiments the segments would have to be prepadded with nans, which is equivalent to stacking the observations that start at t=p (dropping initial observations).

In the AR case this would have the same parameter estimates as SARIMAX except for small differences that come in small samples from different treatment of initial conditions.
One disadvantage of using OLS is that time series features like h-step ahead forecasting is not automatically available and needs to be coded directly.

Aside: A trial dummy as extra exog would allow the mean of the series to vary across experiments, if the assumption that all trials have the same mean is to strong.

(Another way to work around missing values is to impute something finite (e.g. mean) and add an extra dummy variable for each observation/time period with a missing value. The observation specific dummy would remove the effect of this observation on the estimate of other parameter.)

**Comment by gabrielibagon** [comment]
@josef-pkt Thanks very much for the thorough response! 

I'd like to clarify that there is no missing data inside any of the trials. I selectively sliced a sparse collection of small windows from a very long recording. Each trial is also of equal length, and the time series within each trial is synchronized to an external event, so they all have a somewhat similar waveform/autocorrelations. The original context or order of the trials is not important, so they may be shuffled, etc. Here is an illustration of the original recording and the slices that I take from it, where 'x' is a large number of discarded values:

|xxxxxxx|---trial 1---|xxxxxxx|---trial 2---|xxxxxxx|---trial 3---|xxxxxxx|

I would like to use a fitted AR model to take timepoints (y_{t-p} to y_{t-1}) of a new, exogenous trial and forecast the subsequent values of that same trial (y_t to y_{t+n}), where the beginning timepoint of this trial is considered t=0, and I'm interested in forecasting n steps ahead.

If this is what you were describing, I'm not sure I understand your OLS solution. By "creating a lag matrix of the endog as exog", do you mean create a (Num Trials) x (Num Timepoints) matrix? If so, I do not understand why I prepad with nans or drop any trials (since they all have the "same" starting timepoint and ending timepoint). Lastly, I'm not sure I understand where I would put this matrix within the API - in the `AR()` model creation?  Excuse me if I'm misunderstanding this.

I will look more into the state-space models that you mentioned. I think considering each trial to be a full "season" may make more intuitive sense to me.

**Comment by josef-pkt** [comment]
IIRC, this is a paper that I skimmed that makes the point of converting a time series model into a sequence of stacked observations that can be used with standard non-time series methods
https://robjhyndman.com/publications/cv-time-series/

What I meant is to use sm.OLS(endog, exog) instead of AR
for each segment trial get the 
`exog_i = np.column_stack([ y[k :  T - p + k] for k in range(1, p+1)])`
this would have p columns, and T-p rows, we loos p rows because the first observations don't have lagged values available.
(we have a tsatools lagmat helper function for this)
exog_all.append(exog_i)   for all segments i
then stack on top of each others
exog = np.concatenate([exog_all], axis=0)
exog = np.column_stack((exog, exog_other))
endog = <stack all corresponding y_i in a 1-D array>
`sm.OLS(endog, exog).fit().params`

but forecasting h-step will require that we build a new exog each time which includes predicted y for h>1.

The problem for the example is that AR/ARMA are mainly good for short term forecasting. In the long term every stationary AR/ARMA converges to the mean. So I guess in this application it will be important to include the experimental event data, e.g. if there are different types of events or including the lag effects of specific events by including a variable or dummy for time since event. (literature is "intervention analysis", IIRC)
AR/ARMA alone would assume that the response to shocks/noise would be the same for random fluctuations and for the response to the event.

**Comment by ChadFulton** [comment]
I'm not sure that I understand the goal. Do you want each of these segments to follow an AR(p) model, but where the AR coefficients (and maybe the error variance) are constrained to be equal?

If so, one option would be to set up a custom state space model. This could also be done with a multivariate OLS with linear constraints, or with some effort and dummy variables it could be set up with regular OLS and dummy variables with linear restrictions. But I don't think we have those options available?

**Comment by josef-pkt** [comment]
@ChadFulton The basic structure is a dynamic panel model but without cross-sectional correlation, but with common parameters. We don't have any special model for that. We don't have anything like a full multivariate OLS model yet.
The "obvious" solution is to stack all cases which is easy if there are no lag effects, but we cannot just stitch different time series together, without special treatment of the stitches or transition between segments.

In terms of model structure, it would essentially be a separate statespace/tsa model for each segment but with common parameter constraint, but we don't have an easy way to combine those separate models into a single optimization problem with a single set of parameters,
i.e. make a list of submodels, one for each segment, and maximize loglike which is the sum of all submodel loglike. A custom LikelihoodModel for aggregating loglike, score, ... would work.

**Comment by ChadFulton** [comment]
I guess something like this might work or else be a starting point?

http://nbviewer.jupyter.org/gist/ChadFulton/583020ea3740f91bec308212d48a12db

**Comment by noname34** [comment]
> I have a univariate 2D dataset made up of many non-consecutive trials (or slices/windows) of a time series (one trial is not continuous with another trial). I would like to fit an AR model to this collection of trials for forecasting on an exogenous trial.
> 
> My first approach is to concatenate all of the trials into a long 1-dimensional array, padding with NaNs to separate the trials, and then use `AR(missing='drop')` to ignore the NaNs. However, I'm not sure how the `missing` parameter works - does "drop" simply remove the NaNs and concatenate the array on both sides? This is not what I would want, since the AR model would then be trained on the data as if it was continuous, which it is not. The points between the trials shouldn't have any correlation, and the model should not use data from one trial to predict the time series of a new trial.
> 
> Is there a better way to train a univariate AR model on a collection of discontinuous windows of data? Thanks!


@gabrielibagon Have you found a solution to your problem? I have got exactly the same problem. Thank you in advance

**Comment by bashtage** [comment]
If you just want parallel ARs and not MAs, then you can just do the following. Suppose `y` is your data stored as a `DataFrame`:

```
import statsmodels.api as sm
import pandas as pd

p = 3 # AR order
blocks = []
for i in range(len(y)):
    x = y.iloc[:,i]
    block = pd.concat([y.shift(j) for j in range(p+1)], axis=1)
    block.columns = ["lead"] + [f"lag{j} for j in range(1,p+1))]
    block = block.dropna(how="any")
    blocks.append(block)
joined= pd.concat(blocks, axis=0)
lhs = joined.iloc[:,0]
rhs = sm.add_constant(joined.iloc[:,1:])
res = sm.OLS(lhs,rhs).fit()
```

**Comment by noname34** [comment]
Thank you for your quick answer. It is kind of difficult to understand your code because we may not exactly have the same problem. Do you mind helping me on the issue I have just opened? https://github.com/statsmodels/statsmodels/issues/6861

Thank you very much!



Kevin Bürgisser
